File: wathefni-orchestrator/candidates_stage_contract.py
"""Candidates stage-bucket contract — shared status sets for filter expansion proofs.

Mirrors apps/wathefni-dashboard/src/lib/candidatesStageContract.ts.

Axis separation:
* candidates.view.* — job/governance (Talent Pool = no job assigned).
* candidates.stage.* — application lifecycle only.

Talent Pool / needs_role / import_review are not Stage New.
Legacy offered / offer_sent belong to Shortlisted until Offer is a real stage.
"""

from __future__ import annotations

from typing import Any

STAGE_FILTER_NEW = "awaiting_cv"
STAGE_FILTER_READY = "ready_for_review"
STAGE_FILTER_SHORTLISTED = "shortlisted"
STAGE_FILTER_INTERVIEW = "interview"
STAGE_FILTER_HIRED = "hired"
STAGE_FILTER_REJECTED = "rejected"
STAGE_FILTER_WITHDRAWN = "withdrawn"

STAGE_BUCKET_NEW = "new"
STAGE_BUCKET_READY = "ready_for_review"
STAGE_BUCKET_SHORTLISTED = "shortlisted"
STAGE_BUCKET_INTERVIEW = "interview"
STAGE_BUCKET_HIRED = "hired"
STAGE_BUCKET_REJECTED = "rejected"
STAGE_BUCKET_WITHDRAWN = "withdrawn"
STAGE_BUCKET_ARCHIVED = "archived"
STAGE_BUCKET_NONE = "none"
STAGE_BUCKET_UNKNOWN = "unknown"

CANDIDATE_STAGE_FILTER_STATUSES: dict[str, tuple[str, ...]] = {
    STAGE_FILTER_NEW: (
        "awaiting_cv",
        "cv_processing",
        "cv_received",
        "screening",
        "cv_request",
        "cv_upload",
    ),
    STAGE_FILTER_READY: ("ready_for_review", "screening_complete", "review_pending"),
    STAGE_FILTER_SHORTLISTED: ("shortlisted", "offered", "offer_sent"),
    STAGE_FILTER_INTERVIEW: ("interview", "scheduled"),
    STAGE_FILTER_HIRED: ("hired",),
    STAGE_FILTER_REJECTED: ("rejected",),
    STAGE_FILTER_WITHDRAWN: ("withdrawn",),
}

HELD_ACTIVE_STATUSES = frozenset({"needs_role", "import_review"})


def expand_stage_filter_statuses(filter_value: str) -> list[str]:
    key = str(filter_value or "").strip().lower()
    if not key:
        return []
    direct = CANDIDATE_STAGE_FILTER_STATUSES.get(key)
    if direct:
        return list(direct)
    for statuses in CANDIDATE_STAGE_FILTER_STATUSES.values():
        if key in statuses:
            return list(statuses)
    return [key]


def lifecycle_bucket_for_status(status: Any) -> str | None:
    raw = str(status or "").strip().lower()
    if not raw:
        return None
    if raw in CANDIDATE_STAGE_FILTER_STATUSES[STAGE_FILTER_NEW]:
        return STAGE_BUCKET_NEW
    if raw in CANDIDATE_STAGE_FILTER_STATUSES[STAGE_FILTER_READY]:
        return STAGE_BUCKET_READY
    if raw in CANDIDATE_STAGE_FILTER_STATUSES[STAGE_FILTER_SHORTLISTED]:
        return STAGE_BUCKET_SHORTLISTED
    if raw in CANDIDATE_STAGE_FILTER_STATUSES[STAGE_FILTER_INTERVIEW]:
        return STAGE_BUCKET_INTERVIEW
    if raw == STAGE_FILTER_HIRED:
        return STAGE_BUCKET_HIRED
    if raw == STAGE_FILTER_REJECTED:
        return STAGE_BUCKET_REJECTED
    if raw == STAGE_FILTER_WITHDRAWN:
        return STAGE_BUCKET_WITHDRAWN
    if raw == "archived":
        return STAGE_BUCKET_ARCHIVED
    return None


def is_talent_pool_stage_exempt(row: dict[str, Any] | None) -> bool:
    row = row or {}
    status = str(row.get("status") or "").strip().lower()
    return str(row.get("record_state") or "").strip().lower() == "talent_pool" or status in HELD_ACTIVE_STATUSES


def display_stage_bucket(row: dict[str, Any] | None) -> str:
    """Mirror of frontend candidateListStageBucket for live proofs."""
    row = row or {}
    status = str(row.get("status") or "").strip().lower()
    record_state = str(row.get("record_state") or "").strip().lower()
    if record_state == "archived" or status == "import_archived":
        return STAGE_BUCKET_ARCHIVED
    if is_talent_pool_stage_exempt(row):
        return STAGE_BUCKET_NONE
    for key in (status, str(row.get("canonical_stage") or "").strip().lower()):
        bucket = lifecycle_bucket_for_status(key)
        if bucket:
            return bucket
    return STAGE_BUCKET_UNKNOWN

File: apps/wathefni-dashboard/src/lib/candidatesStageContract.ts
/**
 * Candidates stage-bucket contract — single source of truth for list Stage display
 * and Stage filter expansion.
 *
 * Axis separation (approved):
 * - candidates.view.* — job/governance (All, With a job, No job / Talent Pool, Hired,
 *   Archived, Restricted). Owned by unified view predicates.
 * - candidates.stage.* — application lifecycle only.
 *
 * Talent Pool / needs_role / import_review are NOT Stage New.
 * Legacy offered / offer_sent display and filter as Shortlisted until Offer is a
 * real lifecycle stage end-to-end.
 */

/** Toolbar value for label "New" (awaiting_cv kept for saved-view compat). */
export const STAGE_FILTER_NEW = 'awaiting_cv'
export const STAGE_FILTER_READY = 'ready_for_review'
export const STAGE_FILTER_SHORTLISTED = 'shortlisted'
export const STAGE_FILTER_INTERVIEW = 'interview'
export const STAGE_FILTER_HIRED = 'hired'
export const STAGE_FILTER_REJECTED = 'rejected'
export const STAGE_FILTER_WITHDRAWN = 'withdrawn'

/** Display bucket keys (Stage column). */
export const STAGE_BUCKET_NEW = 'new'
export const STAGE_BUCKET_READY = 'ready_for_review'
export const STAGE_BUCKET_SHORTLISTED = 'shortlisted'
export const STAGE_BUCKET_INTERVIEW = 'interview'
export const STAGE_BUCKET_HIRED = 'hired'
export const STAGE_BUCKET_REJECTED = 'rejected'
export const STAGE_BUCKET_WITHDRAWN = 'withdrawn'
export const STAGE_BUCKET_ARCHIVED = 'archived'
/** Quiet non-lifecycle value for Talent Pool / no-job held rows. */
export const STAGE_BUCKET_NONE = 'none'
/** Unmapped lifecycle status — never silently New. */
export const STAGE_BUCKET_UNKNOWN = 'unknown'

export const CANDIDATE_STAGE_FILTER_STATUSES: Record<string, readonly string[]> = {
  [STAGE_FILTER_NEW]: [
    'awaiting_cv',
    'cv_processing',
    'cv_received',
    'screening',
    'cv_request',
    'cv_upload',
  ],
  [STAGE_FILTER_READY]: ['ready_for_review', 'screening_complete', 'review_pending'],
  [STAGE_FILTER_SHORTLISTED]: ['shortlisted', 'offered', 'offer_sent'],
  [STAGE_FILTER_INTERVIEW]: ['interview', 'scheduled'],
  [STAGE_FILTER_HIRED]: ['hired'],
  [STAGE_FILTER_REJECTED]: ['rejected'],
  [STAGE_FILTER_WITHDRAWN]: ['withdrawn'],
}

/** Map a raw application.status (or alias) to a lifecycle display bucket, or null. */
export function lifecycleBucketForStatus(status: string | null | undefined): string | null {
  const raw = String(status || '').trim().toLowerCase()
  if (!raw) return null
  if (CANDIDATE_STAGE_FILTER_STATUSES[STAGE_FILTER_NEW].includes(raw)) return STAGE_BUCKET_NEW
  if (CANDIDATE_STAGE_FILTER_STATUSES[STAGE_FILTER_READY].includes(raw)) return STAGE_BUCKET_READY
  if (CANDIDATE_STAGE_FILTER_STATUSES[STAGE_FILTER_SHORTLISTED].includes(raw)) return STAGE_BUCKET_SHORTLISTED
  if (CANDIDATE_STAGE_FILTER_STATUSES[STAGE_FILTER_INTERVIEW].includes(raw)) return STAGE_BUCKET_INTERVIEW
  if (raw === STAGE_FILTER_HIRED) return STAGE_BUCKET_HIRED
  if (raw === STAGE_FILTER_REJECTED) return STAGE_BUCKET_REJECTED
  if (raw === STAGE_FILTER_WITHDRAWN) return STAGE_BUCKET_WITHDRAWN
  if (raw === 'archived') return STAGE_BUCKET_ARCHIVED
  return null
}

/** Expand a Stage toolbar value (or raw alias) to every matching application.status. */
export function expandStageFilterStatuses(filterValue: string): string[] {
  const key = String(filterValue || '').trim().toLowerCase()
  if (!key) return []
  const direct = CANDIDATE_STAGE_FILTER_STATUSES[key]
  if (direct) return [...direct]
  for (const statuses of Object.values(CANDIDATE_STAGE_FILTER_STATUSES)) {
    if (statuses.includes(key)) return [...statuses]
  }
  return [key]
}

export function isTalentPoolStageExempt(application: {
  is_held?: boolean | null
  record_state?: string | null
  status?: string | null
}): boolean {
  const status = String(application.status || '').trim().toLowerCase()
  return Boolean(
    application.record_state === 'talent_pool'
      || status === 'needs_role'
      || status === 'import_review',
  )
}
